lib/backend/watchersyncer/watchercache.go

```diff
@@ -117,30 +114,15 @@ mainLoop:
 				kvp.Value = nil
 				wc.handleWatchListEvent(kvp)
 			case api.WatchError:
-				// Handle a WatchError.  First determine if the error type indicates that the
-				// watch has closed, and if so we'll need to resync and create a new watcher.
-				wc.results <- event.Error
-
-				if e, ok := event.Error.(cerrors.ErrorWatchTerminated); ok {
-					wc.logger.Debug("Received watch terminated error - recreate watcher")
-					if !e.ClosedByRemote {
-						// If the watcher was not closed by remote, reset the currentWatchRevision.  This will
-						// trigger a full resync rather than simply trying to watch from the last event
-						// revision.
-						wc.logger.Debug("Watch was not closed by remote - full resync required")
-						wc.currentWatchRevision = ""
-						wc.onError()
-					}
-					wc.resyncAndCreateWatcher(ctx)
-				} else {
-					wc.onError()
-				}
+				// Handle a WatchError. This error triggered from upstream, all type
+				// of WatchError are treated equally,log the Error and trigger a full resync.
+
+				wc.logger.WithField("EventType", event.Type).Errorf("Watch error received from Upstream")
+				wc.hasSynced = false
```

Comment:
Might be clearer to replace lines 121 and 122 with a call to onWaitForDatastore().

I think it's not strictly necessary because I think hasSynced is always true at this point - but it means you don't have to think about it :-)

Reply:
Thanks for catching this, I missed this one in the previous commit. Updated PR.